<launch>
	<!-- Doc: http://wiki.ros.org/amcl?distro=kinetic -->

	<arg name="namespace" default="" />

	<!-- Args to set the initial orientation of AMCL node. -->
	<arg name="init_x" default="0.0" />
	<arg name="init_y" default="0.0" />
	<arg name="init_yaw" default="0.0" /> <!-- Radians -->

	<!-- Topic remappings-->
	<arg name="scan_topic" default="base_scan" />
  	<arg name="odom_link" default="odom" />
  	<arg name="base_link" default="base_link" />

	<!-- AMCL Node -->
	<node pkg="amcl" type="amcl" name="amcl" if="$(eval not namespace)" output="screen">
		<!-- Set initial pose -->
		<param name="initial_pose_x" value="$(arg init_x)" />
		<param name="initial_pose_y" value="$(arg init_y)" />
		<param name="initial_pose_a" value="$(arg init_yaw)" />
		<!-- Remap scan topic -->
		<remap from="scan" to="$(arg scan_topic)"/>
		<!-- Remap map topic -->
		<remap from="map" to="/map"/>
		<!-- Remap map service call -->
		<remap from="static_map" to="/static_map" />
		<!-- If not using map_server to provide map, uncomment line below-->
		<!-- <param name="use_map_topic" value="true" /> -->
		<!-- Remap the following topics and services to include a amcl prefix-->
		<remap from="global_localization" to="amcl/global_localization" />
		<remap from="request_nomotion_update" to="amcl/request_nomotion_update" />
		<remap from="set_map" to="amcl/set_map" />
		<remap from="particlecloud" to="amcl/particlecloud" />
		<remap from="amcl_pose" to="amcl/amcl_pose" />
		<remap from="initialpose" to="amcl/initialpose" />

		<!-- Odom frame -->
		<param name="odom_frame_id" value="$(arg odom_link)"/>
		<!-- Base frame -->
		<param name="base_frame_id" value="$(arg base_link)" />

		<!-- Publish scans from best pose at a max of 10 Hz -->
		<param name="odom_model_type" value="diff"/>
		<param name="odom_alpha5" value="0.1"/>
		<param name="transform_tolerance" value="0.2" />
		<param name="gui_publish_rate" value="-1.0"/>
		<param name="laser_max_beams" value="30"/>
		<param name="min_particles" value="100"/>
		<param name="max_particles" value="5000"/>
		<param name="kld_err" value="0.05"/>
		<param name="kld_z" value="0.99"/>
		<param name="odom_alpha1" value="0.2"/>
		<param name="odom_alpha2" value="0.2"/>
		<!-- translation std dev, m -->
		<param name="odom_alpha3" value="0.8"/>
		<param name="odom_alpha4" value="0.2"/>
		<param name="laser_z_hit" value="0.5"/>
		<param name="laser_z_short" value="0.05"/>
		<param name="laser_z_max" value="0.05"/>
		<param name="laser_z_rand" value="0.5"/>
		<param name="laser_sigma_hit" value="0.2"/>
		<param name="laser_lambda_short" value="0.1"/>
		<param name="laser_lambda_short" value="0.1"/>
		<param name="laser_model_type" value="likelihood_field"/>
		<!-- <param name="laser_model_type" value="beam"/> -->
		<param name="laser_likelihood_max_dist" value="2.0"/>
		<param name="update_min_d" value="0.2"/>
		<param name="update_min_a" value="0.5"/>
		<param name="resample_interval" value="1"/>
		<param name="recovery_alpha_slow" value="0.0"/>
		<param name="recovery_alpha_fast" value="0.0"/>
	</node>

	<!-- AMCL Node with namespace-->
	<node pkg="amcl" type="amcl" name="amcl" ns="$(arg namespace)" unless="$(eval not namespace)">
		<!-- Set initial pose -->
		<param name="initial_pose_x" value="$(arg init_x)" />
		<param name="initial_pose_y" value="$(arg init_y)" />
		<param name="initial_pose_a" value="$(arg init_yaw)" />
		<!-- Remap scan topic -->
		<remap from="scan" to="$(arg scan_topic)"/>
		<!-- Remap map topic -->
		<remap from="map" to="/map"/>
		<!-- Remap map service call -->
		<remap from="static_map" to="/static_map" />
		<!-- If not using map_server to provide map, uncomment line below-->
		<!-- <param name="use_map_topic" value="true" /> -->
		<!-- Remap the following topics and services to include a amcl prefix-->
		<remap from="global_localization" to="amcl/global_localization" />
		<remap from="request_nomotion_update" to="amcl/request_nomotion_update" />
		<remap from="set_map" to="amcl/set_map" />
		<remap from="particlecloud" to="amcl/particlecloud" />
		<remap from="amcl_pose" to="amcl/amcl_pose" />
		<remap from="initialpose" to="amcl/initialpose" />

		<!-- Odom frame -->
		<param name="odom_frame_id" value="$(arg odom_link)"/>
		<!-- Base frame -->
		<param name="base_frame_id" value="$(arg base_link)" />

		<!-- Publish scans from best pose at a max of 10 Hz -->
		<param name="odom_model_type" value="diff"/>
		<param name="odom_alpha5" value="0.1"/>
		<param name="transform_tolerance" value="0.2" />
		<param name="gui_publish_rate" value="-1.0"/>
		<param name="laser_max_beams" value="30"/>
		<param name="min_particles" value="100"/>
		<param name="max_particles" value="5000"/>
		<param name="kld_err" value="0.05"/>
		<param name="kld_z" value="0.99"/>
		<param name="odom_alpha1" value="0.2"/>
		<param name="odom_alpha2" value="0.2"/>
		<!-- translation std dev, m -->
		<param name="odom_alpha3" value="0.8"/>
		<param name="odom_alpha4" value="0.2"/>
		<param name="laser_z_hit" value="0.5"/>
		<param name="laser_z_short" value="0.05"/>
		<param name="laser_z_max" value="0.05"/>
		<param name="laser_z_rand" value="0.5"/>
		<param name="laser_sigma_hit" value="0.2"/>
		<param name="laser_lambda_short" value="0.1"/>
		<param name="laser_lambda_short" value="0.1"/>
		<param name="laser_model_type" value="likelihood_field"/>
		<!-- <param name="laser_model_type" value="beam"/> -->
		<param name="laser_likelihood_max_dist" value="2.0"/>
		<param name="update_min_d" value="0.2"/>
		<param name="update_min_a" value="0.5"/>
		<param name="resample_interval" value="1"/>
		<param name="recovery_alpha_slow" value="0.0"/>
		<param name="recovery_alpha_fast" value="0.0"/>
	</node>
</launch>
